### PR TITLE
Oneof case

### PR DESCRIPTION
**Pull request by subash-a**
Creating a oneof case POC.

### CODE REVIEW

File: src/immut_helloworld.ts [hunk 1]
```diff
@@ -2,75 +2,137 @@ import {Uint64} from "./custom_types";
...
+
+    get Email(): string {
+        return this.underlying.getEmail();
+    }
+
+    Serialize(): ArrayBuffer {
+        return this.underlying.serializeBinary();
+    }
+
+    ContactFieldSet(): PersonContactEnum {
```
Comment (myitcv) [inline src/immut_helloworld.ts]: `ContactFieldCase`

File: src/immut_helloworld.ts [hunk 2]
```diff
@@ -74,3 +86,100 @@ export class HelloReply {
...
+		if (this.underlying.getMobile() === num.toNumber()) {
+			return this;
+		} else {
+			let instance = Person.fromPerson(this.underlying.cloneMessage());
+			instance.underlying.setMobile(num.toNumber());
+			return instance;
+		}
+	}
+
+	/* This logic handles the clearing of field in the oneof case */
```
Comment (subash-a) [inline src/immut_helloworld.ts:176]: And provides clear semantic separation between normal fields and fields from a `oneof` group

File: src/immut_helloworld.ts [hunk 3]
```diff
@@ -74,3 +86,100 @@ export class HelloReply {
...
+			return this;
+		} else {
+			let instance = Person.fromPerson(this.underlying.cloneMessage());
+			instance.underlying.setEmail(e === undefined || e === null ? undefined : e);
+			return instance;
+		}
+	}
+
+	ClearEmail(): Person {
+		if (this.underlying.getEmail() === undefined || this.underlying.getEmail() === null) {
```
Comment (myitcv) [inline src/immut_helloworld.ts]: Let's get a separate test in this PR that verifies that a `get` on an 'nil`field is always`undefined`

File: src/helloworld.spec.ts [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+
+			beforeEach(() => {
+				person = new helloworld.Person();
+			});
+
+			it("should have a known starting state", () => {
+				expect(person.getName()).toEqual("");
+				expect(person.getEmail()).not.toBeDefined();
+				expect(person.getMobile()).not.toBeDefined();
+				expect(person.getContactCase()).toEqual(0);
```
Comment (myitcv) [inline src/helloworld.spec.ts]: Use the enum here...

File: src/immut_helloworld.ts [hunk 5]
```diff
@@ -74,3 +86,108 @@ export class HelloReply {
 		return this.underlying.serializeBinary();
 	}
 }
+
+export enum PersonContactEnum {
```
Comment (myitcv) [inline src/immut_helloworld.ts]: Move to JS type definition file

Comment (myitcv) [inline src/immut_helloworld.ts]: ... and then re-export within here

File: src/immut_helloworld.spec.ts [hunk 6]
```diff
@@ -0,0 +1,323 @@
...
+
+describe("HelloRequest", () => {
+
+	describe("New HelloRequest", () => {
+		it("should create a new instance without throwing error", () => {
+			expect(() => new HelloRequest()).not.toThrow();
+		});
+	});
+
+	describe("with a new instance of HelloRequest", () => {
```
Comment (myitcv) [inline src/immut_helloworld.spec.ts]: - Immutable-esque behaviour; we get a new object, don't get a new object if value hasn't changed etc....
- Default values are what we expect
- Setting of `undefined` etc behaves as we expect

**Comment by myitcv** [comment]
Closing as we've covered this elsewhere

**Pull request closed without merging**